llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -7792,6 +7582,29 @@ VectorizationFactor LoopVectorizationPlanner::computeBestVF() {
   VectorizationFactor LegacyVF = selectVectorizationFactor();
   VPlan &BestPlan = getPlanFor(BestFactor.Width);
 
+  // VPlan calculates register pressure from the plan, so it can come to
+  // different conclusions than the legacy cost model.
+  bool RegUsageDeterminedVF = false;
+  if (BestFactor.Width != LegacyVF.Width) {
+    SmallVector<ElementCount, 1> LegacyVFs = {LegacyVF.Width};
+    SmallVector<ElementCount, 1> VFs = {BestFactor.Width};
+
+    auto LegacyRUs =
+        calculateRegisterUsage(getPlanFor(LegacyVF.Width), LegacyVFs, TTI);
+    auto RUs = calculateRegisterUsage(BestPlan, VFs, TTI);
+
+    auto GetMaxUsage = [](SmallMapVector<unsigned, unsigned, 4> MaxLocalUsers) {
+      unsigned Max = 0;
+      for (auto Pair : MaxLocalUsers)
+        if (Pair.second > Max)
+          Max = Pair.second;
+      return Max;
+    };
+    unsigned MaxLegacyRegUsage = GetMaxUsage(LegacyRUs[0].MaxLocalUsers);
+    unsigned MaxRegUsage = GetMaxUsage(RUs[0].MaxLocalUsers);
+    RegUsageDeterminedVF = MaxRegUsage <= MaxLegacyRegUsage;
```

Comment:
My understanding is that the cost model would make a different choice based on register pressure if and only if the legacy reg-pressure calculation for the chosen VF would exceed the number of available registers.

But because the original cost calculation code is not around anymore (this PR removes it), we still need to figure out why the VFs chosen by the legacy/Vplan cost-models don't match. This code attempts to do that by seeing if the LegacyVF would have a higher register usage than the chosen VF. If so, then I'm not sure this is calculating the same thing.

Should this instead add a flag to `calculateRegisterPressure` to have a 'legacy' mode, where it doesn't scale the register usage for a PHI/VPReduction node using the scaling factor? And then you can compare the register usage in the old/new mode, to give the conclusive answer.

Reply:
I had a closer look into this, and ignore my suggestion above. Instead, I think it is sufficient to check that the LegacyVF that would otherwise have been chosen has a register usage that exceeds the maximum number of registers. Otherwise, the cost would have chosen the same VF.

That means, this can be simplified to:
```suggestion
    auto &MLU = LegacyRUs[0].MaxLocalUsers;
    RegUsageDeterminedVF = any_of(MLU, [&](decltype(MLU.front()) &LU) {
      return LU.second > TTI.getNumberOfRegisters(LU.first);
    });
```

Reply:
I've managed to get rid of these changes by checking the register pressure as part of the legacy cost model's VF selection, which mirrors the current behaviour. This has had the knock-on effect of changing some debug output for lots of x86 tests since the plan is used for reg pressure checks, rather than the scalar loop.